src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
     public void add(T value) {
+        Node<T> addedNode = new Node<>(value);
+        if (head == null) {
+            head = tail = addedNode;
+        } else {
+            addLastNode(addedNode);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, size + 1);
+        Node<T> addedNode = new Node<>(value);
+        if (head == null) {
+            head = tail = addedNode;
+        } else if (index == 0) {
+            addFirstNode(addedNode);
+        } else if (index == size) {
+            addLastNode(addedNode);
+        } else {
+            addMiddleNode(index, addedNode);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size);
+        return getCurrentNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, size);
+        Node<T> currentNode = getCurrentNode(index);
+        T result = currentNode.value;
+        currentNode.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size);
+        T result = get(index);
+        unlink(index);
+        size--;
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
+        if (size == 0) {
+            return false;
+        }
+        Node<T> currentNode = head;
+        for (int index = 0; index < size; index++) {
+            if (areEqual(currentNode.value, object)) {
+                unlink(index);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addLastNode(Node<T> addedNode) {
+        tail.next = addedNode;
+        addedNode.prev = tail;
+        tail = addedNode;
+    }
+
+    private void checkIndex(int index, int bound) {
+        if (index < 0 || index >= bound) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of bounds");
+        }
+    }
+
+    private void addFirstNode(Node<T> addedNode) {
+        head.prev = addedNode;
+        addedNode.next = head;
+        addedNode.prev = null;
+        head = addedNode;
+    }
+
+    private void addMiddleNode(int index, Node<T> addedNode) {
+        Node<T> currentNode = getCurrentNode(index);
+        addedNode.next = currentNode;
+        addedNode.prev = currentNode.prev;
+        currentNode.prev.next = addedNode;
+        currentNode.prev = addedNode;
+    }
+
+    private Node<T> getCurrentNode(int index) {
```

Comment:
[Don’t iterate your nodes from the head to tail or vice versa.](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#dont-iterate-your-nodes-from-the-head-to-tail-or-vice-versa)